zngur-def/src/merge.rs

```diff
@@ -0,0 +1,242 @@
+        smallvec.push(item);
+    }
+}
+
+/// Writes the union of `other` and `smallvec` to the latter in O(N * M) time.
+fn inplace_union<T: Eq>(other: Vec<T>, smallvec: &mut std::vec::Vec<T>) {
+    for item in other {
+        push_unique(item, smallvec);
+    }
+}
+
+impl<T: Merge> Merge for Option<T> {
+    /// Writes the partial union of `self` and `into` to the latter.
+    ///
+    /// If both `self` and `into` are Some, the underlying values are merged.
+    /// Otherwise, the result is whichever value is Some, or None if neither is.
+    fn merge(self, into: &mut Self) -> MergeResult {
+        match self {
+            Some(src) => match into.as_mut() {
+                Some(dst) => src.merge(dst),
+                None => {
+                    *into = Some(src);
+                    Ok(())
+                }
+            },
+            None => Ok(()),
+        }
+    }
+}
+
+impl<K, V, I: IntoIterator<Item = (K, V)>> Merge<std::collections::HashMap<K, V>> for I
+where
+    K: Eq + std::hash::Hash,
+    V: Merge,
+{
+    /// Merges a sequence of key-value pairs into a hash map.
+    ///
+    /// If a key is present in both `self` and `into`, the corresponding values are merged.
+    /// Otherwise, the entry from `self` is inserted into `into`.
+    ///
+    /// This implementation implies `std::collections::HashMap<K,V>` is `Merge` for all `V: Merge`,
+    /// because HashMap is `IntoIterator`. We use `IntoIterator` to allow literal sequences of
+    /// key-value pairs to be merged into a map.
+    fn merge(self, into: &mut std::collections::HashMap<K, V>) -> MergeResult {
+        for (key, value) in self {
+            match into.entry(key) {
+                std::collections::hash_map::Entry::Vacant(e) => {
+                    e.insert(value);
+                }
+                std::collections::hash_map::Entry::Occupied(mut e) => {
+                    match value.merge(e.get_mut()) {
+                        Ok(()) => {}
+                        Err(message) => {
+                            return Err(message);
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Merge for ZngurType {
+    /// Writes the partial union of `self` and `into` to the latter.
+    ///
+    /// PRECONDITION: `self.ty == into.ty`.
+    fn merge(self, into: &mut Self) -> MergeResult {
+        if self.ty != into.ty {
+            panic!(
+                "Attempt to merge different types: {} and {}",
+                self.ty, into.ty
+            );
+        }
+
+        if self.layout != into.layout {
+            return Err(MergeFailure::Conflict(
+                "Duplicate layout policy found".to_string(),
+            ));
+        }
+
+        if self.cpp_ref.is_some() && into.layout != LayoutPolicy::ZERO_SIZED_TYPE {
+            return Err(MergeFailure::Conflict(
+                "cpp_ref implies a zero sized stack allocated type".to_string(),
```

Comment:
This isn't a merge error. Is it added in this PR or was it there before? Is there a better place for checking it?

Reply:
So it is now checked in two places? Can we defer this check to the post merged file and unify the checks? It is more a semantic error than a merge error, and we may want to add more of these semantic checks in the future, having them in two places makes things a little hard to maintain.

Reply:
I think we should definitely reduce the duplication here. `merge.rs` seems like a pragmatic place to put these types of checks (rather than deferring the check to the post merged file) because here we'll have an easier time referencing back to the .zng files that produced these semantic entities. That means we'll be able to produce error messages that show both of the source code locations that caused the conflict.

Do you think this is required for this PR? I think this might be better handled in a follow-up PR, where I/we can look at improving the parser error messages?

Reply:
> Do you think this is required for this PR?

No, I think a `TODO` comment would suffice.